### PR TITLE
[DEVOPS-1259] Issue buffer update with multiple projects in parallel

### PR DESCRIPTION
**Pull request by taksan**

### CODE REVIEW

File: core/src/main/java/objective/taskboard/issueBuffer/IssueBufferService.java [hunk 1]
```diff
@@ -8,15 +8,18 @@
 import java.util.Collection;
 import java.util.Date;
 import java.util.HashSet;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/issueBuffer/IssueBufferService.java]: MINOR Remove this unused import 'java.util.concurrent.atomic.AtomicBoolean'. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core/src/main/java/objective/taskboard/issueBuffer/IssueBufferService.java [hunk 2]
```diff
@@ -149,6 +127,67 @@ public void updateIssueBuffer() {
...
+        if (failedProjects.size() > 0) 
+            throw new IllegalStateException("Failed to fetch issues for the following projects: " + 
+                    StringUtils.join(failedProjects, ','));
+        
+        processIssuesWithoutParent(pendingIssues);
+    }
+
+    private void processIssuesWithoutParent(List<IssueScratch> pendingIssues) {
+        IssueScratchResolver resolver = new IssueScratchResolver(issueConverter, parentProvider, this);
+        pendingIssues.forEach(scratch -> resolver.updateIssue(scratch));
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/issueBuffer/IssueBufferService.java]: MINOR Replace this lambda with a method reference. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1612)

File: core/src/main/java/objective/taskboard/issueBuffer/IssueBufferService.java [hunk 3]
```diff
@@ -149,6 +127,67 @@ public void updateIssueBuffer() {
         });
     }
 
+    private void updatesOnNewProjects() {
+        List<IssueScratch> pendingIssues = new LinkedList<>();
+        List<String> failedProjects = new LinkedList<String>();
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/issueBuffer/IssueBufferService.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2293)

File: core/src/main/java/objective/taskboard/issueBuffer/IssueBufferService.java [hunk 4]
```diff
@@ -149,6 +127,67 @@ public void updateIssueBuffer() {
...
+                log.warn("Exception when updating project " + jiraProjectKey + " in issue buffer - Errors:\n" + message, ex);
+                cardsRepo.evictProject(jiraProjectKey);
+                failedProjects.add(jiraProjectKey);
+            } catch (Exception ex) {
+                log.warn("Exception when updating project " + jiraProjectKey + " in issue buffer", ex);
+                cardsRepo.evictProject(jiraProjectKey);
+                failedProjects.add(jiraProjectKey);
+            }
+        });
+        if (failedProjects.size() > 0)
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/issueBuffer/IssueBufferService.java]: MINOR Use isEmpty() to check whether the collection is empty or not. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1155)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: core/src/test/java/objective/taskboard/issueBuffer/IssueBufferServiceTest.java [hunk 5]
```diff
@@ -342,4 +484,193 @@ private IssueUpdateType getIssueUpdateType(final WebhookEvent webhookEvent) {
...
+                        parent = new JSONObject(
+                                "{\n" + 
+                                "    \"key\": \""+parentKey+"\"\n" + 
+                                "}") ;
+                    } catch (JSONException e) {
+                        throw new RuntimeException(e);
+                    }
+                    return parent;
+                }
+                public JiraIssueDto getMock(JiraProjectDto jiraProjectDto) {
```
Comment (herbert-venancio) [inline core/src/test/java/objective/taskboard/issueBuffer/IssueBufferServiceTest.java]: Suggestion:
```Java
                private JsonBuilder.JsonObjectBuilder parentJsonObject() {
                    return JsonBuilder.object(JsonBuilder.property("key", parentKey));
                }

                public JiraIssueDto getMock(JsonBuilder.JsonObjectBuilder jiraProjectDto) {
                    JsonBuilder.JsonObjectBuilder fields = JsonBuilder.object(
                            JsonBuilder.property("project", jiraProjectDto)
                    );
                    if (parentKey != null) {
                        fields.add(
                                JsonBuilder.property("parent", parentJsonObject())
                        );
                    }
                    
                    return JsonBuilder.object(
                            JsonBuilder.property("key", issueKey)
                            , JsonBuilder.property("fields", fields)
                    ).as(JiraIssueDto.class);
                }
```

File: core/src/test/java/objective/taskboard/issueBuffer/IssueBufferServiceTest.java [hunk 6]
```diff
@@ -342,4 +484,193 @@ private IssueUpdateType getIssueUpdateType(final WebhookEvent webhookEvent) {
...
+                    failingProjects.add(projectKey);
+                    return WhenDSL.this;
+                }
+            }
+        }
+        
+        private class ThenDSL {
+            public ThenDSL bufferHasProjects(String ...projects) {
+                
+                assertEquals(StringUtils.join(projects, ","), StringUtils.join(subject.getLoadedProjects(),","));
```
Comment (herbert-venancio) [inline core/src/test/java/objective/taskboard/issueBuffer/IssueBufferServiceTest.java]: Suggestion:
```Java
            public ThenDSL bufferHasNoProject() {
                assertThat(subject.getLoadedProjects()).isEmpty();
                return this;
            }
            public ThenDSL bufferHasProjects(String ...projects) {
                assertThat(subject.getLoadedProjects()).containsExactlyInAnyOrder(projects);
                return this;
            }
```

File: core/src/test/java/objective/taskboard/issueBuffer/IssueBufferServiceTest.java [hunk 7]
```diff
@@ -342,4 +484,193 @@ private IssueUpdateType getIssueUpdateType(final WebhookEvent webhookEvent) {
...
+                return this;
+            }
+
+            public ThenDSL bufferHasIssues(String ...issueList) {
+                String actual = subject.getAllIssues().stream().map(Issue::getIssueKey).collect(Collectors.joining(","));
+                
+                assertEquals(StringUtils.join(issueList,","), actual);
```
Comment (herbert-venancio) [inline core/src/test/java/objective/taskboard/issueBuffer/IssueBufferServiceTest.java]: Suggestion:
```Java
            public ThenDSL bufferHasNoIssues() {
                assertThat(subject.getAllIssues())
                        .isEmpty();
                return this;
            }
            public ThenDSL bufferHasIssues(String ...issueList) {
                assertThat(subject.getAllIssues())
                        .extracting(IssueScratch::getIssueKey)
                        .containsExactlyInAnyOrder(issueList);
                return this;
            }
```

File: test-jira-stub/src/test/java/objective/taskboard/jira/JiraSearchServiceTest.java [hunk 8]
```diff
@@ -194,11 +199,37 @@ public void whenNetworkFailedSearchForOneJql_thenDontSearchLotsOfIssues() {
...
+        Set<JiraIssueDto> allIssues = Collections.synchronizedSet(new LinkedHashSet<>());
         assertThatThrownBy(() -> subject.searchIssues("project = TEST", allIssues::add))
                 .isInstanceOf(RetrofitError.class);
 
         // then
         assertThat(allIssues).size().isLessThan(5000);
     }
+    
+    @Test
+    public void whenNetworkFailsDueToIntermitency_willRetry() {
```
Comment (herbert-venancio) [inline test-jira-stub/src/test/java/objective/taskboard/jira/JiraSearchServiceTest.java]: Typo:
- Intermit[t]ency

File: test-jira-stub/src/test/java/objective/taskboard/jira/JiraSearchResponseBuilder.java [hunk 9]
```diff
@@ -57,6 +64,20 @@ public JiraSearchResponseBuilder startAt(int startAt) {
         return this;
     }
 
+    private static final class SequenceSearchResponseBuilder extends JiraSearchResponseBuilder {
+        private Queue<JiraSearchResponseBuilder> builders = new LinkedList<JiraSearchResponseBuilder>();
```
Comment (herbert-venancio) [inline test-jira-stub/src/test/java/objective/taskboard/jira/JiraSearchResponseBuilder.java]: Use "diamond" notation: `new LinkedList<>();`

File: core/src/main/java/objective/taskboard/jira/JiraSearchService.java [hunk 10]
```diff
@@ -106,22 +109,38 @@ public SearchPageFetcher(String jql, SearchIssueVisitor visitor, Set<String> fie
...
+            for (int i = 0; i < properties.getJiraSearch().getSearchRetryCount()+1; i++)
+                try {
+                    return service.search(buildSearchRequest);
+                }catch(RetrofitError ex) {
+                    log.warn("Failed to execute JQL {} lower bound {}. Will retry", jql, lowerBound);
+                    finalException = ex;
+                    sleeper.sleepUninterruptibly(properties.getJiraSearch().getSearchRetryDelay());
+                }
+            log.error("Giving up executing JQL {} lower bound {}. Permanent Communication failure", jql, lowerBound);
+            throw finalException;
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/jira/JiraSearchService.java]: MAJOR A "NullPointerException" could be thrown; "finalException" is nullable here. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2259)

File: core/src/main/java/objective/taskboard/issueBuffer/CardRepo.java [hunk 11]
```diff
@@ -4,7 +4,9 @@
 import java.util.Collections;
 import java.util.Date;
 import java.util.HashSet;
+import java.util.Iterator;
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/issueBuffer/CardRepo.java]: MINOR Remove this unused import 'java.util.Iterator'. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core/src/main/java/objective/taskboard/issueBuffer/CardRepo.java [hunk 12]
```diff
@@ -4,7 +4,9 @@
...
 import java.util.Map;
+import java.util.Map.Entry;
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/issueBuffer/CardRepo.java]: MINOR Remove this unused import 'java.util.Map.Entry'. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**